jax/_src/scipy/stats/bootstrap.py

```diff
@@ -0,0 +1,240 @@
+  # bca interval only works on 1-sample statistic
+  sample = data[0]
+
+  # calculate z0_hat
+  resample_n = theta_hat_b.shape[-1]
+  theta_hat = statistic(sample, axis=axis)[..., None]
+  percentile = (theta_hat_b < theta_hat).sum(axis=-1) / resample_n
+  z0_hat = ndtri(percentile)
+
+  # calculate a_hat
+  theta_hat_i = _jackknife_resample_and_compute_statistic(sample, statistic)
+  theta_hat_dot = theta_hat_i.mean(axis=-1, keepdims=True)
+  num = ((theta_hat_dot - theta_hat_i) ** 3).sum(axis=-1)
+  den = 6 * ((theta_hat_dot - theta_hat_i) ** 2).sum(axis=-1) ** (3 / 2)
+  a_hat = num / den
+
+  # calculate alpha_1, alpha_2
+  z_alpha = ndtri(alpha)
+  z_1alpha = -z_alpha
+  num1 = z0_hat + z_alpha
+  alpha_1 = ndtr(z0_hat + num1 / (1 - a_hat * num1))
+  num2 = z0_hat + z_1alpha
+  alpha_2 = ndtr(z0_hat + num2 / (1 - a_hat * num2))
+  return alpha_1, alpha_2
+
+
+def _percentile_along_axis(theta_hat_b, alpha):
+  shape = theta_hat_b.shape[:-1]
+  alpha = jnp.broadcast_to(alpha, shape)
+  vmap_percentile = jnp.percentile
+  for i in range(theta_hat_b.ndim - 1):
+    vmap_percentile = vmap(vmap_percentile)
+  percentiles = vmap_percentile(theta_hat_b, alpha)
+  return percentiles[()]
+
+
+def bootstrap_iv(key, data, statistic, vectorized, paired, axis, confidence_level, n_resamples, method):
+  # TODO: add input validations
+  # TODO: handle the case for paired sample
+  # TODO: handle the case for vectorized with arb axis
+  # TODO: handle the case when axis = 0 is used for multi dim data
+  # e.g. data = (np.random.randn(3, 10, 100), ); bootstrap(data, statistic = np.std, axis = 0, confidence_level = 0.95)
+  # how axis argument is used
+  # check alpha is jax array type
+
+
+  if vectorized not in (True, False):
+    raise ValueError("`vectorized` must be `True` or `False`.")
+
+  n_samples = 0
+  try:
+    n_samples = len(data)
+  except TypeError:
+    raise ValueError("`data` must be a sequence of samples.")
+
+  if n_samples == 0:
+    raise ValueError("`data` must contain at least one sample.")
+
+  # enforce data to be jax array type
+  # else will get TracerArrayConversionError in jax.tree_map(lambda data: data[..., idxs], sample)
+  # coz of indexing a numpy array using jax array
+  data_iv = jax.tree_map(lambda data: jnp.asarray(data), data)
+
+  if paired not in (True, False):
+    raise ValueError(f"`paired` must be `True` or `False`.")
+
+  methods = ('percentile', 'basic', 'bca')
+  method = method.lower()
+  if method not in methods:
+    raise ValueError(f"`method` must be in {methods}")
+
+  if not paired and n_samples > 1 and method == 'bca':
+    raise ValueError("`method = 'BCa' is only available for one-sample statistics")
+
+  key, _ = _check_prng_key(key)
+
+  return (key, data_iv, statistic, vectorized, paired, axis,
+          confidence_level, n_resamples, method)
+
+
+@_wraps(
+  scipy.stats.bootstrap,
+  lax_description=_replace_random_state_by_key_no_batch_jnp_statistic_doc,
+  skip_params=("batch",),
```

Comment:
You can use `extra_params` here to document the `key` argument.

That said, I'm starting to wonder if this should really be considered a wrapper of `scipy.stats.bootstrap`, because it's API is now substantially different. In numpy's case, we don't provide any wrappers for `numpy.random` functionality, instead using a different key-based API in `jax.random`. I'm starting to think that the same treatment would make sense here, because as written `jax.scipy.bootstrap` must be called with a different signature than `scipy.bootstrap`.

It also would solve the issue of how to handle irrelevant params like `vectorized`, and we could write the API in a way that is more typical of JAX library functions (i.e. keep batching orthogonal to the implementation, rather than calling `vmap` within.

What do you think?

Reply:
@jakevdp Is there a decision/consensus on whether to adhere to the original API?